trie/smt.py

```diff
@@ -0,0 +1,355 @@
+    Hash32,
+)
+
+from trie.constants import (
+    BLANK_NODE,
+)
+
+from trie.exceptions import (
+    ValidationError,
+)
+
+from trie.validation import (
+    validate_is_bytes,
+    validate_length,
+)
+
+
+def calc_root(key: bytes, value: bytes, branch: Sequence[Hash32]) -> Hash32:
+    """
+    Obtain the merkle root of a given key/value/branch set.
+    Can be used to validate a merkle proof or compute it's value from data.
+
+    :param key: the keypath to decide the ordering of the siblings in the branch
+    :param value: the value (or leaf) that starts the merkle proof computation
+    :param branch: the sequence of siblings used to recursively perform the computation
+
+    :return: the root hash of the merkle proof computation
+
+    .. doctest::
+
+        >>> key = b'\x02'  # Keypath
+        >>> value = b''  # Value (or leaf)
+        >>> branch = tuple([b'\x00'] * 8)  # Any list of hashes
+        >>> calc_root(key, value, branch)
+        b'.+4IKt[\xd2\x14\xe4).\xf5\xc6\n\x11=\x01\xe89\xa1Z\x07#\xfd~(;\xfb\xb8\x8a\x0e'
+
+    """
+    validate_is_bytes(key)
+    validate_is_bytes(value)
+    validate_length(branch, len(key)*8)
+
+    path = to_int(key)
+    target_bit = 1
+    # traverse the path in leaf->root order
+    # branch is in root->leaf order (key is in MSB to LSB order)
+    node_hash = keccak(value)
+    for sibling in reversed(branch):
+        if path & target_bit:
+            node_hash = keccak(sibling + node_hash)
+        else:
+            node_hash = keccak(node_hash + sibling)
+        target_bit <<= 1
+
+    return node_hash
+
+
+class SparseMerkleProof:
+    """
+    Track the current value and merkle proof branch for a given key.
+    This will enable the tracked proof data to stay up to date with changes
+    that may be streamed to the end user over external protocols without having
+    to interactively query the full SMT to obtain the most up-to-date branch.
+
+    Attributes:
+        key: key we are tracking
+        value: currently synchronized value
+        branch: currently synchronized merkle proof branch
+        root_hash: result of computing the merkle root for the tracked data
+
+    .. doctest::
+
+        >>> # smt is located on another process or machine
+        >>> smt = SparseMerkleTree(key_size=1)
+        >>> our_key = b'\x03'
+        >>> our_value = b'\x01'
+        >>> smt.set(our_key, our_value)
+        >>> # We need to track proof data for *some* reason
+        >>> our_proof = SparseMerkleProof(our_key, our_value, smt.branch(our_key))
+        >>> their_key = b'\x05'
+        >>> their_new_value = b'\x01'
+        >>> their_node_updates = smt.set(their_key, their_new_value)
+        >>> # tree updates can be communicated over any channel to proof obj
+        >>> our_proof.merge(their_key, their_new_value, their_node_updates)
+        >>> # Note our branch data was never queried from smt to our proof obj
+        >>> our_proof.branch == smt.branch(our_key)
```

Comment:
> Branch has a particular technical meaning though.

Okay, maybe I'll take back my definition of branch from above, I couldn't find any supporting sources.

> "Siblings" isn't entirely accurate because they are at different levels in the tree.

Sure, in my head it's short for "siblings for the key nodes", or more pedantically "siblings for the nodes corresponding to each bit in the key". I still don't think it's a terrible name, but I'm happy to keep brainstorming.

Other options
- `uncles`
- `ancestor_uncles`
- `pruned_siblings`
- `proof_siblings`
- `proof_nodes`

Reply:
As mentioned above, changed all mentions of sibling to sibling_node. branch is still used to refer to an heirarchically-ordered set of a sibling nodes